REGRESSION(298747@main): [GStreamer] Cannot switch to default audio output device

aperezdc opened the pull request:
#### a8823543d6bf29ae58b360b33d1f86d6267a10e2
<pre>
REGRESSION(298747@main): [GStreamer] Cannot switch to default audio output device
<a href="https://bugs.webkit.org/show_bug.cgi?id=297521">https://bugs.webkit.org/show_bug.cgi?id=297521</a>

Reviewed by Philippe Normand.

When receiving the &quot;default&quot; output device label, search for the
default device in the list of available output devices, or use
the first available one as fallback.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDevice.h:
(WebCore::GStreamerCaptureDevice::device const):
(WebCore::GStreamerCaptureDevice::device): Deleted.

Canonical link: <a href="https://commits.webkit.org/299106@main">https://commits.webkit.org/299106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4274d5d012e3a047e835b3a192dd5d76a3d28f46

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117902 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124046 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69935 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119780 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89468 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120854 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105715 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69964 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29539 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23831 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67710 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99895 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24010 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127129 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44805 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33745 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98138 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101940 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97925 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24907 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43318 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21295 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41234 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44677 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50351 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44137 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47482 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45826 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4622,9 +4622,18 @@ void MediaPlayerPrivateGStreamer::audioOutputDeviceChanged()
     auto deviceId = player->audioOutputDeviceId();
+    if (deviceId == "default"_s) {
+        const auto& devices = GStreamerAudioCaptureDeviceManager::singleton().speakerDevices();
+        if (!devices.isEmpty()) {
```
Maybe tag this as `[[likely]]`?

aperezdc replied:
Done!

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4622,9 +4622,18 @@ void MediaPlayerPrivateGStreamer::audioOutputDeviceChanged()
+                return device.isDefault();
+            });
+            deviceId = (defaultDeviceIndex == notFound) ? devices.first().persistentId() : devices[defaultDeviceIndex].persistentId();
+        }
```
I wonder what to do here in the unlikely case where `devices` is empty...

aperezdc replied:
Currently this will skip setting `deviceId` if `devices` is empty, which means fetching the device from the identifier will fail down below. 

I think we need to change some function types to be able to indicate failures to the rest of the WebKit code, then reject the promise with an error. That is something that eventually we will have to do because [the spec](https://w3c.github.io/mediacapture-output/#dom-htmlmediaelement-setsinkid) covers the following two cases:

- The `Promise` needs to be rejected with a `NotFoundError` if the device identifier passed is not valid, or the device no longer available.
- The `Promise` needs to be rejected with `AbortError` if switching to the chosen device fails.

An empty list of devices would be either case, AFAICS the spec uses the empty string to mean “the default device” (maybe we want to do that in the code as well, as a cleanup), but it says to use `NotFoundError` if the identifier is _not_ the empty string&mdash;then we are left with the second case as possible action.

At any rate, I think these would be improvements better done separately in follow-up patches.

aperezdc committed:
REGRESSION(298747@main): [GStreamer] Cannot switch to default audio output device
https://bugs.webkit.org/show_bug.cgi?id=297521

Reviewed by Philippe Normand.

When receiving the "default" output device label, search for the
default device in the list of available output devices, or use
the first available one as fallback.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDevice.h:
(WebCore::GStreamerCaptureDevice::device const):
(WebCore::GStreamerCaptureDevice::device): Deleted.

Canonical link: https://commits.webkit.org/299106@main
